### PR TITLE
Added Counting Sort [C]

### LINKED ISSUE #164: Counting Sort [C]
http://www.geeksforgeeks.org/counting-sort/

### PR DESCRIPTION
**Pull request by Monal5031**
Fixed #164

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous timespan of 7 days.

### CODE REVIEW

File: counting_sort/counting_sort.c [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        temp[count[arr[i]] - 1] = arr[i];
+        count[arr[i]]--;
+    }
+    for (i = 0; i < n; ++i)
+        arr[i] = temp[i];
+}
+
+int main() {
+	int i;
+	int arr[10] = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
```
Comment (singhpratyush) [inline counting_sort/counting_sort.c]: Better not to define size of `arr`.

File: counting_sort/counting_sort.c [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+*  Time complexity is O(n+k)
+*  Auxiliary space is O(n+k)
+*  n is number of elements and k is the range of input
+*/
+
+#include <stdio.h>
+#include <string.h>
+#define MAX 1000
+
+void counting_sort(int arr[], int n) {
```
Comment (singhpratyush) [inline counting_sort/counting_sort.c]: Add relevant comments.

File: counting_sort/counting_sort.c [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	int temp[n];
+	for (i = 0; i < n; ++i)
+		count[arr[i]]++;
+	for (i = 1; i <= MAX; ++i)
+        count[i] += count[i-1];
+    for (i = 0; i < n; ++i) {
+        temp[count[arr[i]] - 1] = arr[i];
+        count[arr[i]]--;
+    }
+    for (i = 0; i < n; ++i)
```
Comment (Monal5031) [inline counting_sort/counting_sort.c]: Does that work in C? I'm not sure about it.

Comment (Monal5031) [inline counting_sort/counting_sort.c]: Tried that, but it doesn't work.

File: counting_sort/counting_sort.c (hunk 2 above)
Comment (singhpratyush) [inline counting_sort/counting_sort.c]: Please also comment about time complexity.

File: counting_sort/counting_sort.c [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+*/
+
+#include <stdio.h>
+#include <string.h>
+#define MAX 1000
+
+void counting_sort(int arr[], int n) {
+	// Function to sort input array arr.
+	int i;
+	int count[MAX + 1];  // Array to store count's of elements.
```
Comment (aviaryan) [inline counting_sort/counting_sort.c]: It would be better if MAX is fetched from the input rather than fixed.

**Comment by Monal5031** [comment]
@aviaryan Please review.

File: counting_sort/counting_sort.c [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    for (i = 0; i < n; ++i)  // Copy the sorted array back to original array.
+        arr[i] = temp[i];
+}
+
+int main() {
+	int i;
+	int MAX = 10;  // MAX is the maximum number in the array
+	int arr[] = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
+	int n = sizeof(arr) / sizeof(arr[0]);
+	counting_sort(arr, n, MAX);
```
Comment (aviaryan) [inline counting_sort/counting_sort.c]: Full-uppercase variables in C are meant for constants. 
But this isn't a constant as the counting_sort function can be called many times in a program with different values of MAX. 
So please make it `max`.

File: counting_sort/counting_sort.c [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	// Function to sort input array arr.
+	int i;
+	int count[max + 1];  // Array to store count's of elements.
+	memset(count, 0, sizeof(count));
+	int temp[n];
+	for (i = 0; i < n; ++i)  // Count the occurences of contents of array.
+		count[arr[i]]++;
+	for (i = 1; i <= max; ++i)
+        count[i] += count[i-1];
+    for (i = 0; i < n; ++i) {  // Sort the array according to occurences of elements.
```
Comment (aviaryan) [inline counting_sort/counting_sort.c]: Indentation is inconsistent.

**Comment by singhpratyush** [comment]
Don't we need an entry in README?

**Comment by Monal5031** [comment]
@singhpratyush Added the entry to entry.

**Pull request merged**